engine/source_sym_engine.ml

```diff
@@ -0,0 +1,184 @@
+    | x :: xs ->
+       bprintf buf "%a%t%a"
+         bprint x
+         sep
+         (bprint_list ~sep bprint) xs in
+  let break ntabs buf =
+    bprintf buf "\n%s" (BatList.init ntabs (fun _ -> "\t") |> String.concat "") in
+  let rec bprint_tree ntabs buf tree =
+    let sep = break (ntabs+1) in
+    match tree with
+    | Leaf (pilist, expr) ->
+      bprintf buf
+        "Leaf='%a':\
+         %t"
+        bprint_source_expr expr
+        (break (ntabs+1)) ;
+        bprintf buf "constraints {%a}" (bprint_list ~sep bprint_pi) pilist
+    | Node (pi, cst_list) ->
+      bprintf buf
+         "Node: constraints {%a}\
+          %t%a"
+         bprint_pi pi
+         (break (ntabs+1))
+         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
+    | Conjunction (cst_list) ->
+      bprintf buf
+         "Conjunction:%t%a"
+         (break (ntabs+1))
+         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
+    | Root cst_list ->
+       bprintf buf
+         "Root=\
+          %t%a"
+         (break (ntabs+1))
+         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
+  in
+  let buf = Buffer.create 42 in
+  bprint_tree 0 buf stree;
+  BatIO.write_line BatIO.stdout (Buffer.contents buf)
+
+let sym_exec (source: source_program) =
+  let shares_kst ?wildcard_to_none:(w=false) k (clause:clause): clause option =
+    let pattern, expr = clause in 
+    match pattern with
+    | Wildcard when w = false-> Some (Wildcard, expr)
+    | Constructor (k', plist) when k' = k ->
+      if (List.length plist > 1) then
+        Some (Constructor (Tuple, plist), expr)
+      else
+        Some (plist |> List.hd, expr)
+    | _ -> None
+  in
+  let negate_piop = function
+    | EqInt i -> NeqInt i
+    | NeqInt i -> EqInt i
+    | IsK s -> NotK s
+    | NotK s -> IsK s
+    | IsTrue -> IsFalse
+    | IsFalse -> IsTrue
+    | IsTuple -> NotTuple
+    | NotTuple -> IsTuple
+    | IsNil -> NotNil
+    | NotNil -> IsNil
+    | IsCons -> NotCons
+    | NotCons -> IsCons
+    | EqString s -> NeqString s
+    | NeqString s -> EqString s
+  in
+  let piop_of_constructor = function
+    | Variant s -> IsK s
+    | Int i -> EqInt i
+    | Bool b -> if b then IsTrue else IsFalse
+    | String s -> EqString s
+    | Tuple -> IsTuple
+    | Nil -> IsNil
+    | Cons -> IsCons
+  in
+  let rec eval_clauses ?idx:(idx=0) constraints clauses : constraint_tree list =
+    match clauses with
+    | clause::tl -> let pattern, exp = clause
+      in
+      begin match pattern with
+        | Or (p1, p2) ->
+          eval_clauses constraints ((p1,exp)::(p2,exp)::tl) 
+        | Constructor (k, plist) ->
+          let pi = {idx=idx; op=piop_of_constructor k} in
+          let notpi = {pi with op=negate_piop pi.op} in 
+          let constraints' = (notpi :: constraints) in
+          begin match k with
+            | Int _ | String _ | Bool _ ->
+              assert (plist = []);
+              Leaf ([pi], exp)::(eval_clauses constraints' tl)
```

Comment:
I don't think this case is correct. If the clauses look like

```
| 3 -> e1
| p -> e2
```

then this will return a list of the form `([EqInt 3], e1) :: eval_clauses [] [p, e2]`; I would expect the constraint tree for `p` to know that `NeqtInt 3` holds -- otherwise we would have taken the first branch.

My gut feeling is that those constants (and Nil and Cons) below should not be treated differently from the general "Constructor" case. `3` is just one of the (many) constructors (with no arguments) at the `int` type.